wolfcrypt/src/port/Espressif/esp32_mp.c

```diff
@@ -307,143 +682,377 @@ static word32 bits2words(word32 bits)
+    if (ret == MP_OKAY) {
+        ESP_LOGV(TAG, "esp_get_rinv compute success");
+    }
+    else {
         ESP_LOGE(TAG, "failed to calculate mp_mod()");
-        return ret;
     }
 
+#ifdef DEBUG_WOLFSSL
+    if (ret == MP_OKAY) {
+
+        /* computes a = B**n mod b without division or multiplication useful for
+        * normalizing numbers in a Montgomery system. */
+        reti = mp_montgomery_calc_normalization(rinv2, M2);
+        if (reti == MP_OKAY) {
+            ESP_LOGV(TAG, "mp_montgomery_calc_normalization = %d", reti);
+        }
+        else {
+            ESP_LOGW(TAG, "Error montgomery calc M2 result = %d", reti);
+        }
+    }
+#endif
+
+    ESP_LOGV(TAG, "\nEnd esp_get_rinv \n");
     return ret;
 }
+#endif /* ! xEXPTMOD || ! xMULMOD for rinv */
 
-/* Z = X * Y;  */
-int esp_mp_mul(MATH_INT_T* X, MATH_INT_T* Y, MATH_INT_T* Z)
+/* during debug, we'll compare HW to SW results */
+int esp_hw_validation_active(void)
 {
-    int ret;
+#ifdef DEBUG_WOLFSSL
+    return IS_HW_VALIDATION;
+#else
+    return 0; /* we're never validating when not debugging */
+#endif
+}
 
-#ifdef WOLFSSL_SP_INT_NEGATIVE
-    /* neg check: X*Y becomes negative */
-    int neg;
+/* useful during debugging and error display,
+ * we can show all the mp helper calc values */
+int esp_show_mph(struct esp_mp_helper* mph)
+{
+    int ret = MP_OKAY;
+    if (mph->Xs != 0)
+        ESP_LOGI(TAG, "Xs %d", mph->Xs);
+    if (mph->Ys != 0)
+        ESP_LOGI(TAG, "Ys %d", mph->Ys);
+    if (mph->Ms != 0)
+        ESP_LOGI(TAG, "Ms %d", mph->Ms);
+    if (mph->Rs != 0)
+        ESP_LOGI(TAG, "Rs %d", mph->Rs);
+    if (mph->maxWords_sz != 0)
+        ESP_LOGI(TAG, "maxWords_sz %d", mph->maxWords_sz);
+    if (mph->hwWords_sz != 0)
+        ESP_LOGI(TAG, "hwWords_sz %d", mph->hwWords_sz);
+    if (mph->mp != 0)
+        ESP_LOGI(TAG, "mp %d", mph->mp);
+#ifdef DEBUG_WOLFSSL
+    if (mph->mp2 != 0)
+        ESP_LOGI(TAG, "mp2 %d", mph->mp2);
+#endif
+    if (mph->r_inv.used != 0)
+        esp_show_mp("r_inv", &(mph->r_inv));
+    return ret;
+}
 
-    /* aka (X->sign == Y->sign) ? MP_ZPOS : MP_NEG; , but with mp_isneg(): */
-    neg = (mp_isneg(X) == mp_isneg(Y)) ? MP_ZPOS : MP_NEG;
-    if (neg) {
-        /* Negative numbers are relatively infrequent.
-         * May be interesting during verbose debugging: */
-        ESP_LOGV(TAG, "mp_isneg(X) = %d; mp_isneg(Y) = %d; neg = %d ",
-                       mp_isneg(X),      mp_isneg(Y),           neg);
+#if !defined(NO_WOLFSSL_ESP32_CRYPT_RSA_PRI_EXPTMOD) \
+      ||  \
+    !defined(NO_WOLFSSL_ESP32_CRYPT_RSA_PRI_MULMOD)
+/* only when using exptmod or mulmod, we have some helper functions. */
+
+/* given X, Y, M - setup mp hardware and other helper values.*/
+int esp_mp_montgomery_init(MATH_INT_T* X, MATH_INT_T* Y, MATH_INT_T* M,
+                           struct esp_mp_helper* mph)
+{
+    int ret = MP_OKAY;
+    int exp;
+    XMEMSET(mph, 0, sizeof(struct esp_mp_helper));
```

Comment:
does this need null sanity checks on arguments?

Reply:
Probably not needed, only used internally for debugging... but still a good practice. Null check added.